src/sagemaker/s3_utils.py

```diff
@@ -0,0 +1,118 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module contains helper functions related to S3. You may want to use `s3.py` instead.
+
+This has a subset of the functions available through s3.py. This module was initially created with
+functions that were originally in `s3.py` so that those functions could be imported inside
+`session.py` without circular dependencies. (`s3.py` imports Session as a dependency.)
+"""
+from __future__ import print_function, absolute_import
+
+import pathlib
+import logging
+from typing import Optional
+
+from six.moves.urllib.parse import urlparse
+
+logger = logging.getLogger("sagemaker")
+
+
+def parse_s3_url(url):
+    """Returns an (s3 bucket, key name/prefix) tuple from a url with an s3 scheme.
+
+    Args:
+        url (str):
+
+    Returns:
+        tuple: A tuple containing:
+
+            - str: S3 bucket name
+            - str: S3 key
+    """
+    parsed_url = urlparse(url)
+    if parsed_url.scheme != "s3":
+        raise ValueError("Expecting 's3' scheme, got: {} in {}.".format(parsed_url.scheme, url))
+    return parsed_url.netloc, parsed_url.path.lstrip("/")
+
+
+def s3_path_join(*args):
+    """Returns the arguments joined by a slash ("/"), similar to ``os.path.join()`` (on Unix).
+
+    If the first argument is "s3://", then that is preserved.
+
+    Args:
+        *args: The strings to join with a slash. None or empty strings are skipped. For example,
+            an args of ["", "key1", None, "key2"] would yield an output of "key1/key2".
+
+    Returns:
+        str: The joined string, without a slash at the end.
+    """
+    filtered_args = list(filter(lambda item: item is not None and item != "", args))
+    if len(filtered_args) == 0:
+        return ""
+
+    if filtered_args[0].startswith("s3://"):
+        path = str(pathlib.PurePosixPath(*filtered_args[1:])).lstrip("/")
+        return str(pathlib.PurePosixPath(filtered_args[0], path)).replace("s3:/", "s3://")
+
+    return str(pathlib.PurePosixPath(*filtered_args)).lstrip("/")
+
+
+def s3_path_join_with_end_slash(*args):
```

Comment:
Not sure why we are marking this as deprecated and then using it as well. Maybe change wording with recommendation.

Couldnt the `s3_path_join` function handle the extra "/" logic edge case.

Reply:
Ack. I made it for backwards compatibility reasons but wanted to discourage people from using it when adding new code (which is also why I didn't add a param to s3_path_join itself). Will see if I can make clearer